### PR TITLE
Residfp experimental

### PR DESCRIPTION
**Pull request by drfiemost**
Changed the analog stage to use floats instead of ints. The code is clearer but quite heavier on CPU.
Not sure if there's margin for improvements...

### CODE REVIEW

**Comment by reFX-Mike** [comment]
How much of a performance decrease have you measured?

**Comment by drfiemost** [comment]
Near doubled CPU usage, from ~15% to ~27% for a 6581 tune.
Need to profile the code to see where it slows down.

**Comment by reFX-Mike** [comment]
That's quite bad. What CPU did you test with?

**Comment by drfiemost** [comment]
An AMD FX(tm)-6300, it's old but works fine for me :)

**Comment by emoon** [comment]
One thing that may cause a big diff is that `convolve` before was using SIMD code, but now relies on the compiler to generate a good loop (which it may or may not do)

**Comment by reFX-Mike** [comment]
Do you think the better readability of the source is worth the performance cost?

**Comment by reFX-Mike** [comment]
> One thing that may cause a big diff is that `convolve` before was using SIMD code, but now relies on the compiler to generate a good loop (which it may or may not do)

In my case, the compiler (Clang 17.0.6) generated perfect SIMD code, and I had removed the hand-rolled SIMD versions already. Also, even if you replace the convolve function with just a simple "let's average all input values until we have enough", the performance difference was only 10-15% for me. I've briefly toyed with the idea of offering an optional lower-quality version, but the perf-gain was not worth the quality loss.

**Comment by emoon** [comment]
Right, but you have to make sure you have correct flags and such (which may have been missed) for example just `-O3` doesn't generate SIMD here https://godbolt.org/z/9M61qGYnM

**Comment by reFX-Mike** [comment]
Yes, it does. Scroll down a bit more. Clang and MSVC created unrolled loops for even better performance. GCC did not. All of those examples you've linked are SIMD optimized.

**Comment by reFX-Mike** [comment]
Plus, during my tests I've looked at the assembly, because I was confused that the hand rolled SIMD version wasn't any faster than the C++ version. This is when I realised that Clang did such a good job that I didn't need the hand rolled version anymore.

**Comment by emoon** [comment]
This is what I see
```
.LBB0_10:                               # =>This Inner Loop Header: Depth=1
        movss   xmm1, dword ptr [rdi + 4*r8]    # xmm1 = mem[0],zero,zero,zero
        mulss   xmm1, dword ptr [rsi + 4*r8]
        movss   xmm2, dword ptr [rdi + 4*r8 + 4] # xmm2 = mem[0],zero,zero,zero
        addss   xmm1, xmm0
        mulss   xmm2, dword ptr [rsi + 4*r8 + 4]
        addss   xmm2, xmm1
        movss   xmm1, dword ptr [rdi + 4*r8 + 8] # xmm1 = mem[0],zero,zero,zero
        mulss   xmm1, dword ptr [rsi + 4*r8 + 8]
        addss   xmm1, xmm2
        movss   xmm0, dword ptr [rdi + 4*r8 + 12] # xmm0 = mem[0],zero,zero,zero
        mulss   xmm0, dword ptr [rsi + 4*r8 + 12]
        addss   xmm0, xmm1
        add     rcx, -16
        add     r8, 4
        cmp     eax, r8d
        jne     .LBB0_10
```

this uses `mulss` and `addss` which is single value. This loop can be made so `mulps` and `addps` is being used instead (meaning processing 4 values at a time)

**Comment by reFX-Mike** [comment]
Simply add ```-ffast-math``` to the clang options ;)

**Comment by reFX-Mike** [comment]
```.LBB0_6:                                # =>This Inner Loop Header: Depth=1
        movups  xmm2, xmmword ptr [rdi + r11]
        movups  xmm3, xmmword ptr [rdi + r11 + 16]
        movups  xmm4, xmmword ptr [rsi + r11]
        mulps   xmm4, xmm2
        addps   xmm0, xmm4
        movups  xmm2, xmmword ptr [rsi + r11 + 16]
        mulps   xmm2, xmm3
        addps   xmm1, xmm2
        add     r11, 32
        cmp     r10, r11
        jne     .LBB0_6```

**Comment by emoon** [comment]
Which proves my point that if you don't have the correct flags it will not do the above so @drfiemost should double check this.

**Comment by reFX-Mike** [comment]
Sorry, I always assume that people already use the most optimal settings for release builds. It's what I do without even thinking about it.

**Comment by emoon** [comment]
Yeah, so do I, but I have seen many people missing it so I thought it was worth to point out :)

**Comment by drfiemost** [comment]
I don't use `-ffast-math` as it might be harmful in some cases, see:

https://github.com/libsidplayfp/libsidplayfp/blob/febb5ac0675ae0b5d4dab13175db22a9ea332711/src/builders/residfp-builder/residfp/FilterModelConfig6581.cpp#L40-L49

Anyway you're right, the compiler isn't vectorizing the code, I've found a good explanation in [this blog](https://kristerw.github.io/2021/10/19/fast-math/), especially the `-fassociative-math` part. By adding this and related flags the CPU is down to ~24%.
Now it would be cool to force the vectorization from code in some way.

**Comment by reFX-Mike** [comment]
Why not simply use ```std::log1p``` there? You can also wrap only that one function in specific ```--no-fast-math``` clang pragmas.

**Comment by drfiemost** [comment]
> Why not simply use `std::log1p` there? 

Sure, the definition is there only for non-c++11 compilers.


>  You can also wrap only that one function in specific `--no-fast-math` clang pragmas.

Or maybe the opposite, forcing `-ffast-math` for the convolve function so it works event if not defined in the environment.
As a last resort a hand rolled SIMD version could save the day.

**Comment by reFX-Mike** [comment]
> Or maybe the opposite, forcing -ffast-math for the convolve function so it works event if not defined in the environment.

Then, the rest of the code can't benefit from the optimizations.

**Comment by reFX-Mike** [comment]
If the only benefit is marginally better readability, but the cost is a severe decrease in performance (1.6x times slower if we assume ~24%), then it's not worth it.

The improvement in readability could be achieved by different means—local lambdas for common functions.

We also talked about setting C++17 as the minimum, so many possibilities open up which would make the code easier to read. e.g., ```std::clamp``` instead of the typical min/max construct without making these performance sacrifices.

**Comment by drfiemost** [comment]
The problem with integer math is that there's a lot of scaling and shifting which make the code hard to follow, with the added risk of overflows. With floats we can save the clipping for the last stage.
But as it stands the performance drop is too much.

As for vectorization I've found that with `#pragma omp simd` it is possible to force it even without the `fast-math` flag. Also there is the nice `-fopt-info-vec` flag the logs when the compiler actually performs a vectorization.
https://godbolt.org/z/9xG6563hn

**Comment by drfiemost** [comment]
Too bad setting the flag with pragmas is compiler specific. Also I wonder if there's some precision loss involved.
There are a lot of annoying details working with floats :/

I may try again one step at time, starting from the resampler, after reading some docs about float coding tricks.

**Comment by emoon** [comment]
> Also I wonder if there's some precision loss involved.

The order in which you add them together will affect the output, but in the case above I really doubt it would make any difference.

If you "help" the compiler along and write the loop like this
```
float foobar(const float* a, const float* b, int count) {
    float out0 = 0.0f;
    float out1 = 0.0f;
    float out2 = 0.0f;
    float out3 = 0.0f;

    for (int i = 0; i < count/4; ++i) {
        out0 += *a++ * *b++;
        out1 += *a++ * *b++;
        out2 += *a++ * *b++;
        out3 += *a++ * *b++;
    }

    return out0 + out1 + out2 + out3;
}
```
Which is essentially what `-ffast-math` will do. Notice now that even under just regular `-O2` the compiler will generate

```
.L3:
        movups  xmm0, XMMWORD PTR [rsi+rax]
        movups  xmm3, XMMWORD PTR [rdi+rax]
        add     rax, 16
        mulps   xmm0, xmm3
        addps   xmm1, xmm0
        cmp     rdx, rax
        jne     .L3
        movaps  xmm0, xmm1
        movaps  xmm2, xmm1
        shufps  xmm0, xmm1, 85
        addss   xmm0, xmm1
        unpckhps        xmm2, xmm1
        shufps  xmm1, xmm1, 255
        addss   xmm0, xmm2
        addss   xmm0, xmm1
```
So that is of course more code than a single iteration.

This way you can if you want verify the result of a single iteration loop vs the above and see the difference in output, it will be in the very low digits so I doubt there is a problem here.

**Comment by reFX-Mike** [comment]
> The problem with integer math is that there's a lot of scaling and shifting, which make the code hard to follow, with the added risk of overflows.

Why not write a few short helper functions that are always inlined, that do this for you, and always use e.g. int64_t to make sure you get no overflows?

```int64_t mulInt ( const int64_t a, const int64_t b )  {  return ( a * b ) >> 16;  }```
